rest_framework/renderers.py

```diff
@@ -481,35 +481,48 @@ def get_rendered_html_form(self, data, view, method, request):
+            retried = False
+            while True:
+                if existing_serializer is not None:
+                    serializer = existing_serializer
                 else:
-                    # at this point we must have a serializer_class
-                    if method in ('PUT', 'PATCH'):
-                        serializer = self._get_serializer(view.serializer_class, view,
-                                                          request, instance=instance, **kwargs)
+                    if has_serializer:
+                        if method in ('PUT', 'PATCH'):
+                            serializer = view.get_serializer(instance=instance, **kwargs)
+                        else:
+                            serializer = view.get_serializer(**kwargs)
                     else:
-                        serializer = self._get_serializer(view.serializer_class, view,
-                                                          request, **kwargs)
-
-            if hasattr(serializer, 'initial_data'):
-                serializer.is_valid()
-
-            form_renderer = self.form_renderer_class()
-            return form_renderer.render(
-                serializer.data,
-                self.accepted_media_type,
-                dict(
-                    list(self.renderer_context.items()) +
-                    [('template', 'rest_framework/api_form.html')]
-                )
-            )
+                        # at this point we must have a serializer_class
+                        if method in ('PUT', 'PATCH'):
+                            serializer = self._get_serializer(view.serializer_class, view,
+                                                              request, instance=instance, **kwargs)
+                        else:
+                            serializer = self._get_serializer(view.serializer_class, view,
+                                                              request, **kwargs)
+
+                if hasattr(serializer, 'initial_data'):
+                    serializer.is_valid()
+
+                form_renderer = self.form_renderer_class()
+                try:
+                    return form_renderer.render(
+                        serializer.data,
+                        self.accepted_media_type,
+                        dict(
+                            list(self.renderer_context.items()) +
+                            [('template', 'rest_framework/api_form.html')]
+                        )
+                    )
+                except TypeError:
+                    if retried:
+                        # Only retry once
+                        raise
+                    # Try again with a dynamically created serializer and no input.  Having already
+                    # failed with the provided serializer, this is strictly better than a stack
+                    # trace.
+                    existing_serializer = None
+                    kwargs = {}
+                    retried = True
```

Comment:
Which is shorthand for, "couldn't really say without seeing it first".

However, having said that... ...it might be best to *first* make this change in the most minimal way possible, and *only then* consider refactoring if it got accepted.